### ISSUE TITLE
Automatic TileSet collisions based on Terrain

### ISSUE THREAD

**Issue by ChrisMGeo** [godotengine/godot-proposals#6094]
### Describe the project you are working on

2D game with a tilemap

### Describe the problem or limitation you are having in your project

Manually adding tilemap collisions is cumbersome, even with existing and future features like snapping and copying/paste. It'll be faster and way less error-prone to use the painted terrain to generate collisions for selected tiles.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

In `TileSet->Tiles->Select->(select cells you want to generate collisions for)->Physics->Physics Layer` add an option called `Generate Collisions from terrain mapping`
![image](https://user-images.githubusercontent.com/44679968/212440365-b1edb028-50ff-4d73-a9e4-4c139b88590f.png)
Options will be:
```
Terrain Set: int
Terrain: int
Padding: Vector2
```

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Original terrain painting:
![Terrain Painting](https://user-images.githubusercontent.com/44679968/212440693-c88e1e1b-2849-4161-99ae-f3b7b5f83476.png)

Generated collisions with a `Padding` of `(4px, 4px)` (scale not exact):
![Generated Collisions](https://user-images.githubusercontent.com/44679968/212440830-a06fbd86-7bbb-4a05-b97b-e5f41e97b979.png)

I'm assuming that the terrain set uses `Match Corners and Sides`, as I'm unfamiliar with the other modes.

Perhaps this feature can be enhanced in the future to implement rounded or slanted corners, etc.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This can be worked around with an editor plugin, but it's not very convenient.

### Is there a reason why this should be core and not an add-on in the asset library?

This feature should be core because the default `TileSet` collision editor is very inefficient.

**Comment by Fenreliania** [comment]
Improvements to manual collision editing aside, this would be an immensely useful time-saving tool, while also allowing for much more consistency when building a tileset. It makes sense to use the existing information of the terrain flags to also generate collisions, since those things are almost always synonymous anyway.

**Comment by eikobear** [comment]
In my opinion, it makes more sense to allow users to save a TileSet's colliders as a resource and easily reapply them to other sources or TileSets. 

- Trying to generate accurate colliders automatically has a pretty high potential to create sub-par results depending on the design of the tileset. 

- Saving and reusing a tileset collider resource will work for any user as long as their various sources have the same general shape. (And if they don't have the same shape, the generator would have given some of them sub-par results)

- The resource can still have a checkbox to "map colliders by terrain bit" just in case the user has different layouts for different sources

- separately tackle the fact that drawing colliders is a pain. Rather than trying to auto-generate box colliders, we could allow the user to define box colliders using a single drag select rather than placing individual points, or something along those lines

**Comment by Fenreliania** [comment]
> In my opinion, it makes more sense to allow users to save a TileSet's colliders as a resource and easily reapply them to other sources or TileSets.
> 
>     * Saving and reusing a tileset collider resource will work for any user as long as their various sources have the same general shape. (And if they don't have the same shape, the generator would have given some of them sub-par results)

I think this is a great idea also, but doesn't necessarily obviate the usefulness of auto-generated colliders. It's true that usually different tilesets in the same project will generally have the same collisions, but not always. I'm particularly thinking of tilesets that use hard corners vs slopes or stairs. Being able to easily generate colliders with a handful of parameters for size and how corners are handled is still a great time-saving measure that doesn't preclude manual editing, and there's no reason the results would be messy

> 
>     * separately tackle the fact that drawing colliders is a pain. Rather than trying to auto-generate box colliders, we could allow the user to define box colliders using a single drag select rather than placing individual points, or something along those lines

If this was only for generating box colliders I would agree that this is a bit over the top, but I think an effective solution that can generate bevels/slopes and match their intersections across tiles would be very useful.

**Comment by ChrisMGeo** [comment]
> In my opinion, it makes more sense to allow users to save a TileSet's colliders as a resource and easily reapply them to other sources or TileSets.
> 
> * Trying to generate accurate colliders automatically has a pretty high potential to create sub-par results depending on the design of the tileset.
> * Saving and reusing a tileset collider resource will work for any user as long as their various sources have the same general shape. (And if they don't have the same shape, the generator would have given some of them sub-par results)
> * The resource can still have a checkbox to "map colliders by terrain bit" just in case the user has different layouts for different sources
> * separately tackle the fact that drawing colliders is a pain. Rather than trying to auto-generate box colliders, we could allow the user to define box colliders using a single drag select rather than placing individual points, or something along those lines

I agree with the concept of using a separate tileset collider resource, saving them, and reusing them; as well as the improvements to manual collision editing. But I still feel as though this feature can coexist with those and be useful in most cases, as most tilesets have collisions similar to their terrain bitmasks, and then simply use it alongside your mentioned features. Also sorry but I don't quite understand your third point.
